[ci] add gitlab pipeline for criterion benchmarks

sergejparity opened the pull request.

Robbepop commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,49 @@
+  - cargo --version
+  - rustup +nightly show
+  - cargo +nightly --version
+  - cargo spellcheck --version
```
not needed, right?

sergejparity replied:
Yes. It's just debug info. Will remove later

Robbepop replied:
please remove this line for merge

Robbepop reviewed (commented):
looking forward to trying it out. what are your plans on proper formatting of the output?

Robbepop commented:
We might be able to perform the benchmark pass via GitHub Actions using the new hosted GHA runners:
https://github.blog/2022-09-01-github-actions-introducing-the-new-larger-github-hosted-runners-beta/

What do you think?

sergejparity commented:
Yes. Those ones look interesting. But they are still in beta, not sure whether they available for us right now.

paritytech-cicd-pr commented:
| Benchmarks results |
|---|
 Gnuplot not found, using plotters backend
**compile_and_validate/v1**
time:   [5.3728 ms 5.3807 ms 5.3959 ms]
change: [-0.9855% -0.8455% -0.6582%] (p = 0.00 < 0.05)
:white_circle: **Change within noise threshold.**
Found 1 outliers among 10 measurements (10.00%)
1 (10.00%) high severe
||
**instantiate/v1**
time:   [73.992 µs 74.782 µs 75.460 µs]
change: [-0.6179% +0.6305% +2.1037%] (p = 0.41 > 0.05)
:white_circle: **No change in performance detected.**
Found 2 outliers among 10 measurements (20.00%)
1 (10.00%) low severe
1 (10.00%) low mild
||
**execute/tiny_keccak/v1**
time:   [1.2431 ms 1.2468 ms 1.2506 ms]
change: [-0.5937% -0.1709% +0.2057%] (p = 0.46 > 0.05)
:white_circle: **No change in performance detected.**
||
**execute/rev_complement/v1**
execute/rev_complement/v1
time:   [1.3877 ms 1.3880 ms 1.3884 ms]
change: [+0.5521% +0.6682% +0.7440%] (p = 0.00 < 0.05)
:white_circle: **Change within noise threshold.**
||
**execute/regex_redux/v1**
time:   [1.3498 ms 1.3509 ms 1.3528 ms]
change: [-0.6561% -0.3393% -0.0530%] (p = 0.04 < 0.05)
:white_circle: **Change within noise threshold.**
||
**execute/count_until/v1**
time:   [1.8789 ms 1.8791 ms 1.8793 ms]
change: [-2.3138% -1.8681% -1.6168%] (p = 0.00 < 0.05)
:green_circle: **Performance has improved.**
Found 3 outliers among 10 measurements (30.00%)
2 (20.00%) low mild
1 (10.00%) high severe
||
**execute/global_bump/v1**
time:   [3.0445 ms 3.0457 ms 3.0473 ms]
change: [-0.6170% -0.2321% +0.0425%] (p = 0.22 > 0.05)
:white_circle: **No change in performance detected.**
Found 2 outliers among 10 measurements (20.00%)
1 (10.00%) high mild
1 (10.00%) high severe
||
**execute/factorial_recursive/v1**
execute/factorial_recursive/v1
time:   [1.1914 µs 1.1927 µs 1.1948 µs]
change: [+0.1884% +0.3176% +0.4371%] (p = 0.00 < 0.05)
:white_circle: **Change within noise threshold.**
Found 1 outliers among 10 measurements (10.00%)
1 (10.00%) high severe
||
**execute/factorial_iterative/v1**
execute/factorial_iterative/v1
time:   [920.01 ns 920.63 ns 921.42 ns]
change: [-0.0059% +0.0740% +0.1727%] (p = 0.17 > 0.05)
:white_circle: **No change in performance detected.**
Found 2 outliers among 10 measurements (20.00%)
2 (20.00%) high mild
||
**execute/recursive_ok/v1**
time:   [300.80 µs 300.83 µs 300.85 µs]
change: [+7.1133% +7.1277% +7.1421%] (p = 0.00 < 0.05)
:red_circle: **Performance has regressed.**
Found 1 outliers among 10 measurements (10.00%)
1 (10.00%) high mild
||
**execute/recursive_scan/v1**
execute/recursive_scan/v1
time:   [357.66 µs 357.70 µs 357.71 µs]
change: [+6.2929% +6.3627% +6.4161%] (p = 0.00 < 0.05)
:red_circle: **Performance has regressed.**
||
**execute/recursive_trap/v1**
execute/recursive_trap/v1
time:   [26.542 µs 26.559 µs 26.586 µs]
change: [+3.5427% +7.5551% +10.728%] (p = 0.00 < 0.05)
:red_circle: **Performance has regressed.**
Found 2 outliers among 10 measurements (20.00%)
1 (10.00%) low mild
1 (10.00%) high severe
||
**execute/host_calls/v1**
time:   [29.675 µs 30.375 µs 32.172 µs]
change: [-0.3786% +2.2694% +6.9803%] (p = 0.48 > 0.05)
:white_circle: **No change in performance detected.**
Found 1 outliers among 10 measurements (10.00%)
1 (10.00%) high severe
||
**execute/fib_recursive/v1**
execute/fib_recursive/v1
time:   [11.209 ms 11.217 ms 11.231 ms]
change: [+6.1357% +6.2294% +6.3457%] (p = 0.00 < 0.05)
:red_circle: **Performance has regressed.**
Found 1 outliers among 10 measurements (10.00%)
1 (10.00%) high mild
||
**execute/fib_iterative/v1**
execute/fib_iterative/v1
time:   [4.6850 ms 4.7059 ms 4.7329 ms]
change: [-0.2012% +0.1784% +0.6720%] (p = 0.53 > 0.05)
:white_circle: **No change in performance detected.**
Found 2 outliers among 10 measurements (20.00%)
2 (20.00%) high severe
||
**execute/recursive_is_even/v1**
execute/recursive_is_even/v1
time:   [2.1280 ms 2.1352 ms 2.1444 ms]
change: [+0.4736% +0.7173% +0.9758%] (p = 0.00 < 0.05)
:white_circle: **Change within noise threshold.**
||
**execute/memory_sum/v1**
time:   [3.6650 ms 3.6734 ms 3.6792 ms]
change: [-2.5679% -1.9661% -1.3897%] (p = 0.00 < 0.05)
:green_circle: **Performance has improved.**
||
**execute/memory_fill/v1**
time:   [3.9048 ms 3.9049 ms 3.9052 ms]
change: [+0.0509% +0.0596% +0.0684%] (p = 0.00 < 0.05)
:white_circle: **Change within noise threshold.**
||
**execute/memory_vec_add/v1**
execute/memory_vec_add/v1
time:   [7.9111 ms 7.9149 ms 7.9200 ms]
change: [-0.1041% +0.0076% +0.1570%] (p = 0.93 > 0.05)
:white_circle: **No change in performance detected.**
Found 2 outliers among 10 measurements (20.00%)
1 (10.00%) high mild
1 (10.00%) high severe
||

sergejparity commented:
@Robbepop please take a look at the sample report above. Does it look like what is expected?
It was generated from output of `cargo bench --bench benches -- --noplot --baseline master` :
As the base I took what can be seen in the [logs](https://gitlab.parity.io/parity/mirrors/wasmi/-/jobs/1825696)
```
Benchmarking execute/memory_vec_add/v1
Benchmarking execute/memory_vec_add/v1: Warming up for 1.0000 s
Benchmarking execute/memory_vec_add/v1: Collecting 10 samples in estimated 2.1970 s (275 iterations)
Benchmarking execute/memory_vec_add/v1: Analyzing
execute/memory_vec_add/v1
                        time:   [7.9111 ms 7.9149 ms 7.9200 ms]
                        change: [-0.1041% +0.0076% +0.1570%] (p = 0.93 > 0.05)
                        No change in performance detected.
Found 2 outliers among 10 measurements (20.00%)
  1 (10.00%) high mild
  1 (10.00%) high severe
  ```
and transformed into this:
| Benchmarks results |
|---|
**execute/memory_vec_add/v1**
execute/memory_vec_add/v1
time:   [7.9111 ms 7.9149 ms 7.9200 ms]
change: [-0.1041% +0.0076% +0.1570%] (p = 0.93 > 0.05)
:white_circle: **No change in performance detected.**
Found 2 outliers among 10 measurements (20.00%)
1 (10.00%) high mild
1 (10.00%) high severe
||

Any suggestions what needs to be improved, changed, added, etc..?

Robbepop commented:
Hey @sergejparity thank you for the update!
Having those CI based benchmarks is already a big update.
The benchmarked times looks a bit slow, do you know what machine they were running on?
(I did not expect that my old laptop from 2015 would outperform our CI runners :sweat_smile: )

I really like the color encodings where red indicates worse performance, green is improved performance and white is no change!
However, imo the default Criterion output for benchmarks is kind of unreadable at large and I'd much more prefer having a tabular.

Something like the following:

| Benchmark | `master` | PR | Diff % |
|---|---|---|---|
| `compile_and_validate` | 5.3814 ms | 5.3807 ms | :white_circle: -0.8455% |
| `instantiate` | 74.820 µs | 74.782 µs | :white_circle: +0.6305% |
| `execute/tiny_keccak` | 1.2501 ms | 1.2468 ms | :white_circle: -0.1709% |
| `execute/count_until` | 1.9322 ms | 1.8791 ms | :green_circle: -1.8681% |

etc...

Note that I made up the `master` column numbers and took over the PR column numbers as well as the %-diffs.

sergejparity commented:
@Robbepop thank you for the feedback. Will update the template.

What goes regarding performance, I also just using random regular runners for now. Later will tune up this component too.

Robbepop commented:
> @Robbepop thank you for the feedback. Will update the template.
> 
> What goes regarding performance, I also just using random regular runners for now. Later will tune up this component too.

Hyped to see your updates! :)
Glad to hear that runners are going to improve later.

sergejparity commented:
@Robbepop now it should be much closer to what you expect.

Robbepop commented:
@sergejparity yes it looks pretty good to me! Great job! :rocket: 

One small nit is that I'd love if the benchmark names were written in `mono` fonts. I.e. using ` around the names.

Is this PR done?

sergejparity commented:
> names were written in `mono` fonts

Here you go :)

Robbepop commented on scripts/ci/benchmarks-report.sh:
```diff
@@ -0,0 +1,91 @@
+PR_COMMENTS_URL="https://api.github.com/repos/paritytech/wasmi/issues/${CI_COMMIT_BRANCH}/comments"
+
+# master report to json
+echo "PARSING MASTER REPORT"
+sed -e 's/^Found.*//g' \
+    -e 's/^\s\+[[:digit:]].*$//g' \
+    -e 's/\//_/g' \
+    -e 's/^[a-z0-9_]\+/"&": {/g' \
+    -e 's/time:\s\+\[.\{10\}/"time": "/g' \
+    -e 's/.\{10\}\]/"},/g' \
+    -e '1s/^/{\n/g' \
+    -e '/^$/d' \
+    -e 's/  */ /g' \
+    -e 's/^ *\(.*\) *$/\1/' $1 \
+    | sed -z 's/.$//' \
+    | sed -e '$s/.$/}/g' \
+    | tee target/criterion/output_master.json
```
As a follow up to this PR we should try to make this part more readable.
For example criterion supports JSON output as well: https://bheisler.github.io/criterion.rs/book/cargo_criterion/external_tools.html
And with tools such as `jq` we could significantly make extracting information more readable.

I will merge this PR and propose a follow-up PR for you. :) Is that okay with you?

sergejparity replied:
Agree. My homegrown parsing "engine" looks ugly. And definitely can be improved.
But json export tool you mention is a part of `cargo-criterion` which is under development. By the way, have you tried it in action?
Reports from `cargo bench` also produces jsons with raw data in `target/criterion`. I've tried to use them as well, but found that even it is possible get execution times and other metrics. Only thing stopped me from using it is that I forced to write my own statistics analysis tool to interpret them :)
So parsing raw command line output to json appeared easier solution.

sergejparity replied:
Upd: tried to install and use `cargo-criterion` and first attempt was not successful:
```
Benchmarking compile_and_validate/v1thread 'main' panicked at 'could not open benchmark file benches/wasm/wasm_kernel/target/wasm32-unknown-unknown/release/wasm_kernel.wasm: No such file or directory (os er',  wasmi_v1/benches/bench/mod.rs : 16 : 33 
 note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Benchmarking compile_and_validate/v1: Warming up for 1.0000 sthread 'main' panicked at 'Unexpected message FinishedBenchmarkGroup { group: "compile_and_validate/v1" }', /home/sergej/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-criterion-1.1.0/src/bench_target.rs:306:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })', /home/sergej/.cargo/registry/src/github.com-1ecc6299db9ec823/criterion-0.4.0/src/benchmark_group.rs:380:18
stack backtrace:
```

Double checked `benches/wasm/wasm_kernel/target/wasm32-unknown-unknown/release/wasm_kernel.wasm` is in place. What could be the reason for failure.

Robbepop replied:
The problem with raw command line output compared to the JSON output (or genreally machine readable output) is that the former is subject to change whereas the latter is kinda guaranteed to be stable to not break dependents. Therefore as a long term non-brittle solution we really really want to read output that is intended to be machine readable.

Robbepop replied:
> Upd: tried to install and use `cargo-criterion` and first attempt was not successful:
> 
> ```
> Benchmarking compile_and_validate/v1thread 'main' panicked at 'could not open benchmark file benches/wasm/wasm_kernel/target/wasm32-unknown-unknown/release/wasm_kernel.wasm: No such file or directory (os er',  wasmi_v1/benches/bench/mod.rs : 16 : 33 
>  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> Benchmarking compile_and_validate/v1: Warming up for 1.0000 sthread 'main' panicked at 'Unexpected message FinishedBenchmarkGroup { group: "compile_and_validate/v1" }', /home/sergej/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-criterion-1.1.0/src/bench_target.rs:306:26
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Io(Error { kind: UnexpectedEof, message: "failed to fill whole buffer" })', /home/sergej/.cargo/registry/src/github.com-1ecc6299db9ec823/criterion-0.4.0/src/benchmark_group.rs:380:18
> stack backtrace:
> ```
> 
> Double checked `benches/wasm/wasm_kernel/target/wasm32-unknown-unknown/release/wasm_kernel.wasm` is in place. What could be the reason for failure.

You are probably doing this in the root directory instead of the `wasmi_v1` subdirectory.

sergejparity replied:
You are right. Now it works, at first glance json's produced by `cargo-criterion` are almost what is needed. Of course some processing also is needed to produce readable output, like time conversion from ns into more convenient units. But at least we'll have an out-of-the-box machine readable data

Robbepop commented:
Thanks a lot @sergejparity for this great work! I will merge this PR now and will write some issues for future work items for you and the benchmarking CI. :)

:rocket:
